### DISCUSSION TITLE
Type Hinting the Connection class in the ConnectionPool

### DISCUSSION THREAD

**Discussion by vkcku** [psycopg/psycopg#558] [Q&A] [answered]
Hi. I passed in a custom connection class to the `AsyncConnectionPool` to be able to track the duration of the queries and such, but I was unable to properly type hint that the connection that is returned from `await pool.connection()` is my custom class. I looked at the source code and it seems that generics are not used for `AsyncConnectionPool`. Is this a mistake in my understanding, or is this something that can be included? If it's already possible to properly type hint that, how can I do so?

**Comment by dvarrazzo** [comment]
Maybe it is something that is being addressed in #540, @dlax? Or is your MR only about row factories?

**Reply by dlax** [reply]
No; it's not addressed in #540.

**Reply by dvarrazzo** [reply]
If we had `ConnectionPool[Connection[DictRow]]`, could we fix both the sides of the issue... or am I wrong?

**Reply by dlax** [reply]
#540 adds this, so this brings `pool.connection()` returning a `Connection[Row]` . But it seems to me that providing what's asked here (i.e. returning `MyConnection[Row]` when `ConnectionPool(..., connection_class=MyConnection)`) would require another level of generics which is not possible in Python at the moment.

**Reply by dvarrazzo** [reply, marked as answer]
Let's see... I've tried to make the pool generic on the connection and that seems to work ok - see #559. I would like to see if I can merge this and #540 so that it would be possible to specify both subclass and connection subclass, if the latter is generic.

**Reply by vkcku** [reply]
Thank you very much. I think #559 does what I was hoping for.
